### PR TITLE
Implemented passwords

### PR DESCRIPTION
**Pull request by hmartyscu**

### CODE REVIEW

File: src/main/java/com/example/compound/entities/User.java [hunk 1]
```diff
@@ -18,35 +18,42 @@
...
+     * @param password the User's password
      */
-    public User(String name, double balance, String email) {
-        super(name, balance, email);
-        this.UUID = 0;
-        this.username = email;
-        this.expenses = new ArrayList<>();
+    public User(String name, double balance, String email, String password) {
+        // TODO: Take in a distinct username
+        this(0, name, email, email, balance, password); // TODO: Should this constructor with no UUID ever be used?
```
Comment (WaseeA) [inline src/main/java/com/example/compound/entities/User.java:35]: No, this constructor should not be used in our final program. A user with no UUID will cause issues when dealing with database storage.

File: src/main/java/com/example/compound/entities/User.java [hunk 2]
```diff
@@ -18,35 +18,42 @@
...
+    }
+
+    public void setPassword(String password) { // TODO: Add test in UserTest
+        this.password = password;
     }
 
     @Override
     public String generateUUID() {
-        return null;
+        return null; // TODO: Maybe remove this and have a repository generate a UUID
```
Comment (WaseeA) [inline src/main/java/com/example/compound/entities/User.java:56]: Unclear what you mean by "repository generate a UUID". Do you mean create a function that reads our database to generate a unique UUID?

File: src/main/java/com/example/compound/entities/User.java [hunk 3]
```diff
@@ -57,4 +64,6 @@ public String toString() {
     public void addExpense(Expense E) {
         this.expenses.add(E.getEUID());
     }
+
+    // TODO: removeExpense?
```
Comment (WaseeA) [inline src/main/java/com/example/compound/entities/User.java:68]: For now let us leave it. We need to further discuss how we will redesign "group expenses".

File: src/main/java/com/example/compound/controller/Controller.java [hunk 4]
```diff
@@ -310,7 +312,7 @@ private void caseYHelper(InOut inOut, HashMap<Person, Double> borrowedSoFar, Has
         // Otherwise, create a stand in person.
         else {
             Person standIn = this.userManager.createUser(
-                    name, 0.0, email);
+                    name, 0.0, email, ""); // TODO: PersonManager.createPerson?
```
Comment (WaseeA) [inline src/main/java/com/example/compound/controller/Controller.java:315]: Agreed, we should have that function. Perhaps we can implement a builder pattern and add PremiumUser and other subclasses?

File: src/main/java/com/example/compound/entities/User.java [hunk 5]
```diff
@@ -18,35 +18,42 @@
...
         this.password = password;
+        this.expenses = new ArrayList<>();
+    }
+
+    public String getPassword() { // TODO: Add test in UserTest
+        return password;
+    }
+
+    public void setPassword(String password) { // TODO: Add test in UserTest
+        this.password = password;
```
Comment (WaseeA) [inline src/main/java/com/example/compound/entities/User.java:51]: Maybe we can modify this function to include some type of regex search for security? For example check if there is a number, if it is repeated characters etc.

**Review by WaseeA — COMMENTED** [review]
I've included comments in Controller.java and User.java which would address some questions and brings up further discussion needed.

File: src/main/java/com/example/compound/entities/User.java (hunk 3 above)
Comment (hmartyscu) [reply src/main/java/com/example/compound/entities/User.java:68]: OK. I'll leave it as a TODO for now.

File: src/main/java/com/example/compound/entities/User.java (hunk 2 above)
Comment (hmartyscu) [reply src/main/java/com/example/compound/entities/User.java:56]: Yes. The Data class stores Budgets directly, and in the addBudget method, a new Budget is assigned a BUID:
```
        String BUID = Integer.toString(this.budgetCounter);
        budget.setBUID(BUID);
```
New UIDs will probably be set in a different way with an actual database. Maybe the the database / data access class's method for generating UIDs could be called by a use case class.

File: src/main/java/com/example/compound/entities/User.java (hunk 1 above)
Comment (hmartyscu) [reply src/main/java/com/example/compound/entities/User.java:35]: I created new issues based on this and two other comments of yours.

**Pull request merged**